packages/truffle-external-compile/index.js

```diff
@@ -0,0 +1,154 @@
+  const { cwd, logger, input } = options;
+  const child = exec(command, { cwd, input });
+
+  child.stdout.on('data', function(data) {
+    data = data.toString().replace(/\n$/, '');
+    logger.log(data);
+  });
+
+  child.stderr.on('data', function(data) {
+    data = data.toString().replace(/\n$/, '');
+    logger.log(data);
+  });
+
+  child.on('close', function(code) {
+    // If the command didn't exit properly, show the output and throw.
+    if (code !== 0) {
+      var err = new Error("Unknown exit code: " + code);
+      return callback(err);
+    }
+
+    callback();
+  });
+});
+
+async function processTargets (targets, cwd) {
+  const contracts = {};
+  for (let target of targets) {
+    let targetContracts = await processTarget(target, cwd);
+    for (let [name, contract] of Object.entries(targetContracts)) {
+      contracts[name] = contract;
+    }
+  }
+
+  return contracts;
+}
+
+async function processTarget (target, cwd) {
+  const usesPath = target.path != undefined;
+  const usesCommand = target.command != undefined;
+  const usesStdin = target.stdin || target.stdin == undefined;  // default true
+  const usesProperties = target.properties || target.fileProperties;
+
+  if (usesProperties && usesPath) {
+    throw new Error(
+      "External compilation target cannot define both properties and path"
+    );
+  }
+
+  if (usesProperties && usesCommand) {
+    throw new Error(
+      "External compilation target cannot define both properties and command"
+    );
+  }
+
+
+  if (usesCommand && !usesPath) {
+    // just run command
+    const output = execSync(target.command, { cwd });
+    const contract = JSON.parse(output);
+    return { [contract.contractName]: contract };
+  }
+
+  if (usesPath && !glob.hasMagic(target.path)) {
+    // individual file
+    const filename = resolve(cwd, target.path);
+    let input, command, execOptions;
+    if (usesStdin) {
+      input = fs.readFileSync(filename).toString();
+      command = target.command;
+      execOptions = { cwd, input };
+    } else {
+      command = `${target.command} ${filename}`;
+      execOptions = { cwd };
+    }
+
+    const output = (usesCommand)
+      ? execSync(command, execOptions)
+      : input;
+
+    const contract = JSON.parse(output);
+    return { [contract.contractName]: contract };
+  }
+
+  if (usesPath && glob.hasMagic(target.path)) {
+    // glob expression, recurse after expansion
+    let paths = await glob(target.path, { cwd, follow: true });
+    // copy target properties, overriding path with expanded form
+    let targets = paths.map(path => Object.assign({}, target, { path }));
+    return await processTargets(targets, cwd);
+  }
+
+  if (usesProperties) {
+    // contract properties listed individually
+    const contract = Object.assign({}, target.properties || {});
+
+    for (let [key, path] of Object.entries(target.fileProperties || {})) {
+      const contents = fs.readFileSync(resolve(cwd, path)).toString();
```

Comment:
If you change `toString()` to `toString('hex')` here, it (at least partially) fixes the bytecode issue.

Reply:
Here is a slightly more sophisticated approach: https://github.com/lrettig/truffle/commit/acfc8b35e46b89554df1689d936a7829d068c5e5

Reply:
I don't think that is a good idea. Why make a distinction for wasm?